mythril/laser/ethereum/plugins/mutation_pruner.py

```diff
@@ -0,0 +1,56 @@
+from mythril.laser.ethereum.state.annotation import StateAnnotation
+from mythril.laser.ethereum.svm import LaserEVM
+from mythril.laser.ethereum.plugins.signals import PluginSkipWorldState
+from mythril.laser.ethereum.state.global_state import GlobalState
+from mythril.laser.ethereum.transaction.transaction_models import (
+    ContractCreationTransaction,
+)
+
+
+class MutationAnnotation(StateAnnotation):
+    """Mutation Annotation
+
+    This is the annotation used by the MutationPruner plugin to record mutations
+    """
+
+    pass
+
+
+class MutationPruner:
+    """Mutation pruner plugin
+
+    Let S be a world state from which T is a symbolic transaction, and S' is the resulting world state.
+    In a situation where T does not execute any mutating instructions we can safely abandon further analysis on top of
+    state S'. This is for the reason that we already performed analysis on S, which is equivalent.
+
+    This optimization inhibits path explosion caused by "clean" behaviour
+
+    The basic operation of this plugin is as follows:
+     - Hook all mutating operations and introduce a MutationAnnotation to the global state on execution of the hook
+     - Hook the svm EndTransaction on execution filter the states that do not have a mutation annotation
+
+    """
+
+    def __init__(self):
+        pass
+
+    def initialize(self, symbolic_vm: LaserEVM):
+        """Initializes the mutation pruner
+
+        Introduces hooks for SSTORE operations
+        :param symbolic_vm:
+        :return:
+        """
+
+        @symbolic_vm.pre_hook("SSTORE")
```

Comment:
It's worth noting that call instructions also mutate the contract's state if their value is > 0. We don't accurately model contract balances though (#689), so this doesn't necessarily have to be implemented here immediately, but we should be aware of it.

Reply:
Calls with >0 value do need to be considered mutated even though we don't model balance. Otherwise we can get false negatives on the Ether Thief check with counts Ether sent vs. Ether received over all transactions. At least It think so...

Reply:
There are a few other instructions that mutate the state as well (suicide for example), however I'm not sure if they "matter" for multi-transactional analysis and opening up new possible paths

Reply:
@JoranHonig Why was this marked resolved? Did we decide that calls with >0 value do not need to be considered mutated?

Reply:
I don't think that this is a problem. 
If we have a transaction sequence of 
A -> B -> C
Where C does nothing but put money in the contract, then you find
A -> B -> C -> D 
Where D has the withdrawal. Since the vuln in D is independent of the balance change in C it must also be possible to find:
A -> B -> D
Therefore, if we abandon execution of C we should still find transaction D

Reply:
Ok I added a check so only traces with a constraint callvalue == 0 are ignored

Reply:
Do we also want to hook `CALL` opcodes?